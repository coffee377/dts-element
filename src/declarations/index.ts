export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
