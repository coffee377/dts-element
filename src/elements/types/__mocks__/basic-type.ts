import {Type} from '../../type';

export class BasicType extends Type<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[BasicType ${this.parameters.name}]`;
  }

}
