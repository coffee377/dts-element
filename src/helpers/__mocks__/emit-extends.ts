export const emit_extends = (): string => '[emit-extends]';
