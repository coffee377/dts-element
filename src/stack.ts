import {Element} from './element';
import {instance_of, IConstructor} from './utils/instance-of';

export class Stack {

  public raw_stack: Element[];

  constructor(raw_stack: Element[] = []) {
    this.raw_stack = raw_stack;
  }

  public push(element: Element): Stack {
    return new Stack([...this.raw_stack, element]);
  }

  public instances_of(constructors: IConstructor<Element>[]): boolean {
    return (this.raw_stack.length < constructors.length)
      ? false
      : this.raw_stack
        .slice(0, constructors.length)
        .every(instance_of(constructors));
  }

  public last_instances_of(constructors: IConstructor<Element>[]): boolean {
    return (this.raw_stack.length < constructors.length)
      ? false
      : this.raw_stack
        .slice(-constructors.length)
        .reverse()
        .every(instance_of(constructors));
  }

}
