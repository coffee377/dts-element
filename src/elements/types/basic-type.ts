import {Stack} from '../../stack';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_stack: Stack): string {
    return this.parameters.name;
  }

}
