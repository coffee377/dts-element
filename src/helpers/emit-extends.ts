import {GenericDefinition} from '../elements/generic-definition';
import {Type} from '../elements/type';
import {Stack} from '../stack';

export const emit_extends = (an_extends: Type | null, stack: Stack): string =>
  (an_extends === null) || !(stack.last_instances_of([GenericDefinition]))
    ? ''
    : ` extends ${an_extends.emit(stack)}`;
