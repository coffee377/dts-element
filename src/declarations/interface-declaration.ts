import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {transform} from '../transform';
import {IGeneralType} from '../types/general-type';
import {create_object_type, IObjectType} from '../types/object-type';
import {add_declare_modifier_if_need, create_type_arguments, create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface IInterfaceDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  type?: IObjectType;
  extends?: IGeneralType[];
}

export interface IInterfaceDeclaration
  extends IElement<ElementKind.InterfaceDeclaration>, IInterfaceDeclarationOptions {}

export const create_interface_declaration = (options: IInterfaceDeclarationOptions): IInterfaceDeclaration => ({
  ...create_element(ElementKind.InterfaceDeclaration),
  ...options,
});

// tslint:disable:ter-indent max-line-length

export const transform_interface_declaration = (element: IInterfaceDeclaration, path: IElement<any>[]) =>
  ts.createInterfaceDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
                            (element.export === true)
                              ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
                              : undefined,
                            path,
                          ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ element.extends && [ts.createHeritageClause(
                            /* token */ ts.SyntaxKind.ExtendsKeyword,
                            /* types */ element.extends.map(interface_type => ts.createExpressionWithTypeArguments(
                                          /* typeArguments */ create_type_arguments(interface_type.generics, path),
                                          /* expression    */ ts.createIdentifier(interface_type.name),
                                        )),
                          )],
    /* members         */ (transform(element.type || create_object_type(), path) as ts.TypeLiteralNode).members,
  );
