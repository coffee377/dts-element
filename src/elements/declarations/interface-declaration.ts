import {emit_export} from '../../helpers/emit-export';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_members} from '../../helpers/emit-members';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ObjectMember} from '../members/object-member';
import {GenericType} from '../types/generic-type';

export interface IInterfaceDeclarationRequiredParameters {
  name: string;
}

export interface IInterfaceDeclarationOptionalParameters {
  generics: GenericType[];
  members: ObjectMember[];
  export: boolean;
}

/**
 * ```ts
 * interface some_interface_name {
 *   // some object members
 * }
 * ```
 */
export class InterfaceDeclaration
    extends Declaration<IInterfaceDeclarationRequiredParameters, IInterfaceDeclarationOptionalParameters> {

  private _instance_of_interface_declaration: true;

  public get default_parameters(): IDeclarationOptionalParameters & IInterfaceDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      members: [],
      export: false,
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, members, generics} = this.parameters;
    const generic = emit_generics(generics, stack, true);
    const an_export = emit_export(this.parameters.export, stack);
    return `${an_export}interface ${name}${generic} ${emit_members(members, stack)}`;
  }

}
