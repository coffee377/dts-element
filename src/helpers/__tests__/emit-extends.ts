jest.unmock('../emit-extends');

import {InterfaceDeclaration} from '../../elements/declarations/interface-declaration';
import {BasicType} from '../../elements/types/basic-type';
import {Stack} from '../../stack';
import {emit_extends} from '../emit-extends';

const name = 'A';
const basic_type = new BasicType({name});
const interface_declaration: InterfaceDeclaration = Object.create(InterfaceDeclaration.prototype);

it('should return empty string while extends is null', () => {
  expect(emit_extends(null, new Stack([interface_declaration]))).toBe('');
});

it('should return empty string while stack is not InterfaceDeclaration', () => {
  expect(emit_extends(basic_type, new Stack())).toBe('');
});

it('should return extends with emitted type while extends is not null and stack is InterfaceDeclaration', () => {
  expect(emit_extends(basic_type, new Stack([interface_declaration]))).toMatchSnapshot();
});
