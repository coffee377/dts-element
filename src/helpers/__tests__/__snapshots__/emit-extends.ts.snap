// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return extends with emitted type while extends is not null and stack is GenericDefinition 1`] = `" extends [BasicType A]"`;
